Detect root directory renames

When I rename a directory, Zed doesn't update to reflect the rename.

I'm still able to reproduce this with Zed 0.142:

```sh
mkdir -p /tmp/junk
zed /tmp/junk
mv /tmp/junk /tmp/stuff
```

Zed.log:
```
[ERROR] failed to canonicalize root path: No such file or directory (os error 2)
```

--- a/crates/fs/src/fs.rs
+++ b/crates/fs/src/fs.rs
@@ -9,6 +9,14 @@ use git::GitHostingProviderRegistry;
 
 #[cfg(target_os = "linux")]
 use ashpd::desktop::trash;
+use libc::c_char;
+use libc::F_GETPATH;
+use std::ffi::{CStr, OsStr};
+use std::os::fd::AsFd;
+use std::os::fd::AsRawFd;
+use std::os::fd::IntoRawFd;
+use std::os::fd::RawFd;
+use std::os::unix::ffi::OsStrExt;
 #[cfg(target_os = "linux")]
 use std::{fs::File, os::fd::AsFd};
 
@@ -95,7 +103,7 @@ pub trait Fs: Send + Sync {
     async fn trash_file(&self, path: &Path, options: RemoveOptions) -> Result<()> {
         self.remove_file(path, options).await
     }
-    async fn open_handle(&self, path: &Path) -> Result<i64>;
+    async fn open_handle(&self, path: &Path) -> Result<Arc<dyn FileHandle>>;
     async fn open_sync(&self, path: &Path) -> Result<Box<dyn io::Read>>;
     async fn load(&self, path: &Path) -> Result<String> {
         Ok(String::from_utf8(self.load_bytes(path).await?)?)
@@ -186,6 +194,26 @@ pub struct Metadata {
 pub struct RealFs {
     git_hosting_provider_registry: Arc<GitHostingProviderRegistry>,
     git_binary_path: Option<PathBuf>,
+}
+
+pub trait FileHandle: Send + Sync + std::fmt::Debug {
+    fn current_path(&self) -> Result<PathBuf>;
+}
+
+impl FileHandle for std::fs::File {
+    fn current_path(&self) -> Result<PathBuf> {
+        let fd = self.as_fd();
+        let mut path_buf: [c_char; libc::PATH_MAX as usize] = [0; libc::PATH_MAX as usize];
+
+        let result = unsafe { libc::fcntl(fd.as_raw_fd(), libc::F_GETPATH, path_buf.as_mut_ptr()) };
+        if result == -1 {
+            anyhow::bail!("Failed to get path for file descriptor".to_string());
+        }
+
+        let c_str = unsafe { CStr::from_ptr(path_buf.as_ptr()) };
+        let path = PathBuf::from(OsStr::from_bytes(c_str.to_bytes()));
+        Ok(path)
+    }
 }
 
 pub struct RealWatcher {}
@@ -401,8 +429,8 @@ impl Fs for RealFs {
         Ok(Box::new(std::fs::File::open(path)?))
     }
 
-    async fn open_handl(&self, path: &Path) -> Result<Box<dyn io::Read>> {
-        Ok(Box::new(std::fs::File::open(path)?))
+    async fn open_handle(&self, path: &Path) -> Result<Arc<dyn FileHandle>> {
+        Ok(Arc::new(std::fs::File::open(path)?))
     }
 
     async fn load(&self, path: &Path) -> Result<String> {
@@ -1367,6 +1395,15 @@ impl Watcher for FakeWatcher {
     }
 }
 
+#[derive(Debug)]
+struct FakeHandle;
+
+impl FileHandle for FakeHandle {
+    fn current_path(&self) -> Result<PathBuf> {
+        anyhow::bail!("current_path on FakeHandle not implemented")
+    }
+}
+
 #[cfg(any(test, feature = "test-support"))]
 #[async_trait::async_trait]
 impl Fs for FakeFs {
@@ -1647,6 +1684,14 @@ impl Fs for FakeFs {
     async fn open_sync(&self, path: &Path) -> Result<Box<dyn io::Read>> {
         let bytes = self.load_internal(path).await?;
         Ok(Box::new(io::Cursor::new(bytes)))
+    }
+
+    async fn open_handle(&self, path: &Path) -> Result<Box<dyn FileHandle>> {
+        self.simulate_random_delay().await;
+        let state = self.state.lock();
+        let entry = state.read_path(&path)?;
+        let entry = entry.lock();
+        Ok(Box::new(FakeHandle))
     }
 
     async fn load(&self, path: &Path) -> Result<String> {

--- a/crates/fs/src/fs.rs
+++ b/crates/fs/src/fs.rs
@@ -55,14 +55,14 @@ pub trait Watcher: Send + Sync {
     fn remove(&self, path: &Path) -> Result<()>;
 }
 
-#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
 pub enum PathEventKind {
     Removed,
     Created,
     Changed,
 }
 
-#[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
 pub struct PathEvent {
     pub path: PathBuf,
     pub kind: Option<PathEventKind>,
@@ -974,6 +974,7 @@ impl FakeFs {
                 events_paused: false,
                 read_dir_call_count: 0,
                 metadata_call_count: 0,
+                moves: Default::default(),
             }),
         });
 
@@ -1721,7 +1722,7 @@ impl Fs for FakeFs {
         Ok(Box::new(io::Cursor::new(bytes)))
     }
 
-    async fn open_handle(&self, path: &Path) -> Result<Box<dyn FileHandle>> {
+    async fn open_handle(&self, path: &Path) -> Result<Arc<dyn FileHandle>> {
         self.simulate_random_delay().await;
         let state = self.state.lock();
         let entry = state.read_path(&path)?;
@@ -1731,7 +1732,7 @@ impl Fs for FakeFs {
             FakeFsEntry::Dir { inode, .. } => inode,
             _ => unreachable!(),
         };
-        Ok(Box::new(FakeHandle { inode }))
+        Ok(Arc::new(FakeHandle { inode }))
     }
 
     async fn load(&self, path: &Path) -> Result<String> {

--- a/crates/fs/src/fs.rs
+++ b/crates/fs/src/fs.rs
@@ -1429,7 +1429,7 @@ struct FakeHandle {
 impl FileHandle for FakeHandle {
     fn current_path(&self, fs: &Arc<dyn Fs>) -> Result<PathBuf> {
         let state = fs.as_fake().state.lock();
-        let Some(target) = state.moves.get(&self.inode).clone() else {
+        let Some(target) = state.moves.get(&self.inode) else {
             anyhow::bail!("fake fd not moved")
         };
 

--- a/crates/fs/src/fs.rs
+++ b/crates/fs/src/fs.rs
@@ -95,6 +95,7 @@ pub trait Fs: Send + Sync {
     async fn trash_file(&self, path: &Path, options: RemoveOptions) -> Result<()> {
         self.remove_file(path, options).await
     }
+    async fn open_handle(&self, path: &Path) -> Result<i64>;
     async fn open_sync(&self, path: &Path) -> Result<Box<dyn io::Read>>;
     async fn load(&self, path: &Path) -> Result<String> {
         Ok(String::from_utf8(self.load_bytes(path).await?)?)
@@ -397,6 +398,10 @@ impl Fs for RealFs {
     }
 
     async fn open_sync(&self, path: &Path) -> Result<Box<dyn io::Read>> {
+        Ok(Box::new(std::fs::File::open(path)?))
+    }
+
+    async fn open_handl(&self, path: &Path) -> Result<Box<dyn io::Read>> {
         Ok(Box::new(std::fs::File::open(path)?))
     }
 

--- a/crates/fs/src/fs.rs
+++ b/crates/fs/src/fs.rs
@@ -9,16 +9,9 @@ use git::GitHostingProviderRegistry;
 
 #[cfg(target_os = "linux")]
 use ashpd::desktop::trash;
-use libc::c_char;
-use libc::F_GETPATH;
-use std::ffi::{CStr, OsStr};
+use std::fs::File;
 use std::os::fd::AsFd;
 use std::os::fd::AsRawFd;
-use std::os::fd::IntoRawFd;
-use std::os::fd::RawFd;
-use std::os::unix::ffi::OsStrExt;
-#[cfg(target_os = "linux")]
-use std::{fs::File, os::fd::AsFd};
 
 #[cfg(unix)]
 use std::os::unix::fs::MetadataExt;
@@ -201,18 +194,31 @@ pub trait FileHandle: Send + Sync + std::fmt::Debug {
 }
 
 impl FileHandle for std::fs::File {
+    #[cfg(target_os = "macos")]
     fn current_path(&self) -> Result<PathBuf> {
         let fd = self.as_fd();
         let mut path_buf: [c_char; libc::PATH_MAX as usize] = [0; libc::PATH_MAX as usize];
 
         let result = unsafe { libc::fcntl(fd.as_raw_fd(), libc::F_GETPATH, path_buf.as_mut_ptr()) };
         if result == -1 {
-            anyhow::bail!("Failed to get path for file descriptor".to_string());
+            anyhow::bail!("fcntl returned -1".to_string());
         }
 
-        let c_str = unsafe { CStr::from_ptr(path_buf.as_ptr()) };
-        let path = PathBuf::from(OsStr::from_bytes(c_str.to_bytes()));
+        let c_str = unsafe { std::ffi::CStr::from_ptr(path_buf.as_ptr()) };
+        let path = PathBuf::from(std::ffi::OsStr::from_bytes(c_str.to_bytes()));
         Ok(path)
+    }
+
+    #[cfg(target_os = "linux")]
+    fn current_path(&self) -> Result<PathBuf> {
+        let fd = self.as_fd();
+        let fd_path = format!("/proc/self/fd/{}", fd.as_raw_fd());
+        Ok(std::fs::read_link(fd_path)?)
+    }
+
+    #[cfg(target_os = "windows")]
+    fn current_path(&self) -> Result<PathBuf> {
+        anyhow::bail!("unimplemented")
     }
 }
 
@@ -1395,9 +1401,11 @@ impl Watcher for FakeWatcher {
     }
 }
 
+#[cfg(any(test, feature = "test-support"))]
 #[derive(Debug)]
 struct FakeHandle;
 
+#[cfg(any(test, feature = "test-support"))]
 impl FileHandle for FakeHandle {
     fn current_path(&self) -> Result<PathBuf> {
         anyhow::bail!("current_path on FakeHandle not implemented")

--- a/crates/fs/src/fs.rs
+++ b/crates/fs/src/fs.rs
@@ -9,8 +9,11 @@ use git::GitHostingProviderRegistry;
 
 #[cfg(target_os = "linux")]
 use ashpd::desktop::trash;
+#[cfg(target_os = "linux")]
 use std::fs::File;
+#[cfg(unix)]
 use std::os::fd::AsFd;
+#[cfg(unix)]
 use std::os::fd::AsRawFd;
 
 #[cfg(unix)]
@@ -190,34 +193,39 @@ pub struct RealFs {
 }
 
 pub trait FileHandle: Send + Sync + std::fmt::Debug {
-    fn current_path(&self) -> Result<PathBuf>;
+    fn current_path(&self, fs: &Arc<dyn Fs>) -> Result<PathBuf>;
 }
 
 impl FileHandle for std::fs::File {
     #[cfg(target_os = "macos")]
-    fn current_path(&self) -> Result<PathBuf> {
+    fn current_path(&self, _: &Arc<dyn Fs>) -> Result<PathBuf> {
+        use std::{
+            ffi::{CStr, OsStr},
+            os::unix::ffi::OsStrExt,
+        };
+
         let fd = self.as_fd();
-        let mut path_buf: [c_char; libc::PATH_MAX as usize] = [0; libc::PATH_MAX as usize];
+        let mut path_buf: [libc::c_char; libc::PATH_MAX as usize] = [0; libc::PATH_MAX as usize];
 
         let result = unsafe { libc::fcntl(fd.as_raw_fd(), libc::F_GETPATH, path_buf.as_mut_ptr()) };
         if result == -1 {
             anyhow::bail!("fcntl returned -1".to_string());
         }
 
-        let c_str = unsafe { std::ffi::CStr::from_ptr(path_buf.as_ptr()) };
-        let path = PathBuf::from(std::ffi::OsStr::from_bytes(c_str.to_bytes()));
+        let c_str = unsafe { CStr::from_ptr(path_buf.as_ptr()) };
+        let path = PathBuf::from(OsStr::from_bytes(c_str.to_bytes()));
         Ok(path)
     }
 
     #[cfg(target_os = "linux")]
-    fn current_path(&self) -> Result<PathBuf> {
+    fn current_path(&self, _: &Arc<dyn Fs>) -> Result<PathBuf> {
         let fd = self.as_fd();
         let fd_path = format!("/proc/self/fd/{}", fd.as_raw_fd());
         Ok(std::fs::read_link(fd_path)?)
     }
 
     #[cfg(target_os = "windows")]
-    fn current_path(&self) -> Result<PathBuf> {
+    fn current_path(&self, _: &Arc<dyn Fs>) -> Result<PathBuf> {
         anyhow::bail!("unimplemented")
     }
 }
@@ -794,6 +802,7 @@ struct FakeFsState {
     buffered_events: Vec<PathEvent>,
     metadata_call_count: usize,
     read_dir_call_count: usize,
+    moves: std::collections::HashMap<u64, PathBuf>,
 }
 
 #[cfg(any(test, feature = "test-support"))]
@@ -1403,12 +1412,22 @@ impl Watcher for FakeWatcher {
 
 #[cfg(any(test, feature = "test-support"))]
 #[derive(Debug)]
-struct FakeHandle;
+struct FakeHandle {
+    inode: u64,
+}
 
 #[cfg(any(test, feature = "test-support"))]
 impl FileHandle for FakeHandle {
-    fn current_path(&self) -> Result<PathBuf> {
-        anyhow::bail!("current_path on FakeHandle not implemented")
+    fn current_path(&self, fs: &Arc<dyn Fs>) -> Result<PathBuf> {
+        let state = fs.as_fake().state.lock();
+        let Some(target) = state.moves.get(&self.inode).clone() else {
+            anyhow::bail!("fake fd not moved")
+        };
+
+        if state.try_read_path(&target, false).is_some() {
+            return Ok(target.clone());
+        }
+        anyhow::bail!("fake fd target not found")
     }
 }
 
@@ -1549,6 +1568,14 @@ impl Fs for FakeFs {
                 Err(anyhow!("path does not exist: {}", &old_path.display()))
             }
         })?;
+
+        let inode = match *moved_entry.lock() {
+            FakeFsEntry::File { inode, .. } => inode,
+            FakeFsEntry::Dir { inode, .. } => inode,
+            _ => 0,
+        };
+
+        state.moves.insert(inode, new_path.clone());
 
         state.write_path(&new_path, |e| {
             match e {
@@ -1699,7 +1726,12 @@ impl Fs for FakeFs {
         let state = self.state.lock();
         let entry = state.read_path(&path)?;
         let entry = entry.lock();
-        Ok(Box::new(FakeHandle))
+        let inode = match *entry {
+            FakeFsEntry::File { inode, .. } => inode,
+            FakeFsEntry::Dir { inode, .. } => inode,
+            _ => unreachable!(),
+        };
+        Ok(Box::new(FakeHandle { inode }))
     }
 
     async fn load(&self, path: &Path) -> Result<String> {

--- a/crates/fs/src/fs.rs
+++ b/crates/fs/src/fs.rs
@@ -221,7 +221,15 @@ impl FileHandle for std::fs::File {
     fn current_path(&self, _: &Arc<dyn Fs>) -> Result<PathBuf> {
         let fd = self.as_fd();
         let fd_path = format!("/proc/self/fd/{}", fd.as_raw_fd());
-        Ok(std::fs::read_link(fd_path)?)
+        let new_path = std::fs::read_link(fd_path)?;
+        if new_path
+            .file_name()
+            .is_some_and(|f| f.to_string_lossy().ends_with(" (deleted)"))
+        {
+            anyhow::bail!("file was deleted")
+        };
+
+        Ok(new_path)
     }
 
     #[cfg(target_os = "windows")]

--- a/crates/remote_server/src/remote_editing_tests.rs
+++ b/crates/remote_server/src/remote_editing_tests.rs
@@ -1087,6 +1087,45 @@ async fn test_reconnect(cx: &mut TestAppContext, server_cx: &mut TestAppContext)
 }
 
 #[gpui::test]
+async fn test_remote_root_rename(cx: &mut TestAppContext, server_cx: &mut TestAppContext) {
+    let fs = FakeFs::new(server_cx.executor());
+    fs.insert_tree(
+        "/code",
+        json!({
+            "project1": {
+                ".git": {},
+                "README.md": "# project 1",
+            },
+        }),
+    )
+    .await;
+
+    let (project, _) = init_test(&fs, cx, server_cx).await;
+
+    let (worktree, _) = project
+        .update(cx, |project, cx| {
+            project.find_or_create_worktree("/code/project1", true, cx)
+        })
+        .await
+        .unwrap();
+
+    cx.run_until_parked();
+
+    fs.rename(
+        &PathBuf::from("/code/project1"),
+        &PathBuf::from("/code/project2"),
+        Default::default(),
+    )
+    .await
+    .unwrap();
+
+    cx.run_until_parked();
+    worktree.update(cx, |worktree, cx| {
+        assert_eq!(worktree.root_name(), "project2")
+    })
+}
+
+#[gpui::test]
 async fn test_remote_git_branches(cx: &mut TestAppContext, server_cx: &mut TestAppContext) {
     let fs = FakeFs::new(server_cx.executor());
     fs.insert_tree(

--- a/crates/remote_server/src/remote_editing_tests.rs
+++ b/crates/remote_server/src/remote_editing_tests.rs
@@ -1120,7 +1120,7 @@ async fn test_remote_root_rename(cx: &mut TestAppContext, server_cx: &mut TestAp
     .unwrap();
 
     cx.run_until_parked();
-    worktree.update(cx, |worktree, cx| {
+    worktree.update(cx, |worktree, _| {
         assert_eq!(worktree.root_name(), "project2")
     })
 }

--- a/crates/worktree/src/worktree.rs
+++ b/crates/worktree/src/worktree.rs
@@ -1089,10 +1089,10 @@ impl LocalWorktree {
                             if let Some(new_path) = new_path {
                                 this.snapshot.git_repositories = Default::default();
                                 this.snapshot.ignores_by_parent_abs_path = Default::default();
-                                this.snapshot.root_name = new_path
+                                let root_name = new_path
                                     .file_name()
                                     .map_or(String::new(), |f| f.to_string_lossy().to_string());
-                                this.snapshot.abs_path = new_path;
+                                this.snapshot.update_abs_path(new_path, root_name);
                             }
                             this.restart_background_scanners(cx);
                         }
@@ -2093,11 +2093,23 @@ impl Snapshot {
             .and_then(|entry| entry.git_status)
     }
 
+    fn update_abs_path(&mut self, abs_path: Arc<Path>, root_name: String) {
+        self.abs_path = abs_path;
+        if root_name != self.root_name {
+            self.root_char_bag = root_name.chars().map(|c| c.to_ascii_lowercase()).collect();
+            self.root_name = root_name;
+        }
+    }
+
     pub(crate) fn apply_remote_update(&mut self, mut update: proto::UpdateWorktree) -> Result<()> {
         log::trace!(
             "applying remote worktree update. {} entries updated, {} removed",
             update.updated_entries.len(),
             update.removed_entries.len()
+        );
+        self.update_abs_path(
+            Arc::from(PathBuf::from(update.abs_path).as_path()),
+            update.root_name,
         );
 
         let mut entries_by_path_edits = Vec::new();
@@ -3757,11 +3769,19 @@ impl BackgroundScanner {
                     .lock()
                     .snapshot
                     .root_file_descriptor
-                    .current_path()
+                    .current_path(&self.fs)
                     .log_err()
                     .filter(|new_path| **new_path != *root_path);
 
-                log::error!("failed to canonicalize root path: {}", err);
+                if let Some(new_path) = new_path.as_ref() {
+                    log::info!(
+                        "root renamed from {} to {}",
+                        root_path.display(),
+                        new_path.display()
+                    )
+                } else {
+                    log::warn!("root path could not be canonicalized: {}", err);
+                }
                 self.status_updates_tx
                     .unbounded_send(ScanState::RootUpdated {
                         new_path: new_path.map(|p| p.into()),

--- a/crates/worktree/src/worktree.rs
+++ b/crates/worktree/src/worktree.rs
@@ -345,6 +345,9 @@ enum ScanState {
         barrier: SmallVec<[barrier::Sender; 1]>,
         scanning: bool,
     },
+    RootUpdated {
+        new_path: Option<Arc<Path>>,
+    },
 }
 
 struct UpdateObservationState {
@@ -1082,6 +1085,17 @@ impl LocalWorktree {
                             *this.is_scanning.0.borrow_mut() = scanning;
                             this.set_snapshot(snapshot, changes, cx);
                             drop(barrier);
+                        }
+                        ScanState::RootUpdated { new_path } => {
+                            if let Some(new_path) = new_path {
+                                this.snapshot.git_repositories = Default::default();
+                                this.snapshot.ignores_by_parent_abs_path = Default::default();
+                                this.snapshot.root_name = new_path
+                                    .file_name()
+                                    .map_or(String::new(), |f| f.to_string_lossy().to_string());
+                                this.snapshot.abs_path = new_path;
+                            }
+                            this.restart_background_scanners(cx);
                         }
                     }
                     cx.notify();
@@ -3739,21 +3753,21 @@ impl BackgroundScanner {
         let root_canonical_path = match self.fs.canonicalize(&root_path).await {
             Ok(path) => path,
             Err(err) => {
-                if let Some(new_path) = self
+                let new_path = self
                     .state
                     .lock()
                     .snapshot
                     .root_file_descriptor
                     .current_path()
                     .log_err()
-                    .filter(|new_path| new_path != root_path)
-                {
-                    dbg!("root path changed: {}", new_path);
-                    self.state.lock().snapshot.abs_path = new_path;
-                } else {
-                    log::error!("failed to canonicalize root path: {}", err);
-                    return;
-                }
+                    .filter(|new_path| **new_path != *root_path);
+
+                log::error!("failed to canonicalize root path: {}", err);
+                self.status_updates_tx
+                    .unbounded_send(ScanState::RootUpdated {
+                        new_path: new_path.map(|p| p.into()),
+                    });
+                return;
             }
         };
 

--- a/crates/worktree/src/worktree.rs
+++ b/crates/worktree/src/worktree.rs
@@ -54,6 +54,7 @@ use std::{
     future::Future,
     mem,
     ops::{AddAssign, Deref, DerefMut, Sub},
+    os::fd::RawFd,
     path::{Path, PathBuf},
     pin::Pin,
     sync::{
@@ -287,7 +288,7 @@ pub struct LocalSnapshot {
     git_repositories: TreeMap<ProjectEntryId, LocalRepositoryEntry>,
     /// The file handle of the root dir
     /// (so we can find it after it's been moved)
-    root_file_descriptor: i64,
+    root_file_descriptor: Arc<dyn fs::FileHandle>,
 }
 
 struct BackgroundScannerState {
@@ -385,7 +386,7 @@ impl Worktree {
             true
         });
 
-        let handle = fs.open_sync(path)
+        let root_file_descriptor = fs.open_handle(&abs_path).await?;
 
         cx.new_model(move |cx: &mut ModelContext<Worktree>| {
             let mut snapshot = LocalSnapshot {
@@ -398,6 +399,7 @@ impl Worktree {
                         .map_or(String::new(), |f| f.to_string_lossy().to_string()),
                     abs_path,
                 ),
+                root_file_descriptor,
             };
 
             if let Some(metadata) = metadata {
@@ -3737,8 +3739,21 @@ impl BackgroundScanner {
         let root_canonical_path = match self.fs.canonicalize(&root_path).await {
             Ok(path) => path,
             Err(err) => {
-                log::error!("failed to canonicalize root path: {}", err);
-                return;
+                if let Some(new_path) = self
+                    .state
+                    .lock()
+                    .snapshot
+                    .root_file_descriptor
+                    .current_path()
+                    .log_err()
+                    .filter(|new_path| new_path != root_path)
+                {
+                    dbg!("root path changed: {}", new_path);
+                    self.state.lock().snapshot.abs_path = new_path;
+                } else {
+                    log::error!("failed to canonicalize root path: {}", err);
+                    return;
+                }
             }
         };
 

--- a/crates/worktree/src/worktree.rs
+++ b/crates/worktree/src/worktree.rs
@@ -287,7 +287,7 @@ pub struct LocalSnapshot {
     git_repositories: TreeMap<ProjectEntryId, LocalRepositoryEntry>,
     /// The file handle of the root dir
     /// (so we can find it after it's been moved)
-    root_file_descriptor: Arc<dyn fs::FileHandle>,
+    root_file_handle: Option<Arc<dyn fs::FileHandle>>,
 }
 
 struct BackgroundScannerState {
@@ -388,7 +388,7 @@ impl Worktree {
             true
         });
 
-        let root_file_descriptor = fs.open_handle(&abs_path).await?;
+        let root_file_handle = fs.open_handle(&abs_path).await.log_err();
 
         cx.new_model(move |cx: &mut ModelContext<Worktree>| {
             let mut snapshot = LocalSnapshot {
@@ -401,7 +401,7 @@ impl Worktree {
                         .map_or(String::new(), |f| f.to_string_lossy().to_string()),
                     abs_path,
                 ),
-                root_file_descriptor,
+                root_file_handle,
             };
 
             if let Some(metadata) = metadata {
@@ -3768,9 +3768,9 @@ impl BackgroundScanner {
                     .state
                     .lock()
                     .snapshot
-                    .root_file_descriptor
-                    .current_path(&self.fs)
-                    .log_err()
+                    .root_file_handle
+                    .clone()
+                    .and_then(|handle| handle.current_path(&self.fs).log_err())
                     .filter(|new_path| **new_path != *root_path);
 
                 if let Some(new_path) = new_path.as_ref() {

--- a/crates/worktree/src/worktree.rs
+++ b/crates/worktree/src/worktree.rs
@@ -54,7 +54,6 @@ use std::{
     future::Future,
     mem,
     ops::{AddAssign, Deref, DerefMut, Sub},
-    os::fd::RawFd,
     path::{Path, PathBuf},
     pin::Pin,
     sync::{
@@ -3766,7 +3765,8 @@ impl BackgroundScanner {
                 self.status_updates_tx
                     .unbounded_send(ScanState::RootUpdated {
                         new_path: new_path.map(|p| p.into()),
-                    });
+                    })
+                    .ok();
                 return;
             }
         };

--- a/crates/worktree/src/worktree.rs
+++ b/crates/worktree/src/worktree.rs
@@ -285,6 +285,9 @@ pub struct LocalSnapshot {
     /// All of the git repositories in the worktree, indexed by the project entry
     /// id of their parent directory.
     git_repositories: TreeMap<ProjectEntryId, LocalRepositoryEntry>,
+    /// The file handle of the root dir
+    /// (so we can find it after it's been moved)
+    root_file_descriptor: i64,
 }
 
 struct BackgroundScannerState {
@@ -381,6 +384,8 @@ impl Worktree {
             );
             true
         });
+
+        let handle = fs.open_sync(path)
 
         cx.new_model(move |cx: &mut ModelContext<Worktree>| {
             let mut snapshot = LocalSnapshot {

